### DISCUSSION TITLE
isDirty should align with other libraries

### DISCUSSION THREAD

**Discussion by alexqhj** [TanStack/form#1081] [Ideas] [answered]
`isDirty` flips to `true` when you change any form value, but doesn't return to `false` when the state returns back to default values.
I'm unsure about the reasoning for this behavior. 

Here's an example using Tanstack Form.
https://stackblitz.com/edit/tanstack-form-1cspfar1?file=src%2Findex.tsx

The docs specify:

>Important note for users coming from React Hook Form: the isDirty flag in TanStack/form is different from the flag with the same name in RHF. In RHF, isDirty = true, when the form's values are different from the original values. If the user changes the values in a form, and then changes them again to end up with values that match the form's default values, isDirty will be false in RHF, but true in TanStack/form. The default values are exposed both on the form's and the field's level in TanStack/form (form.options.defaultValues, field.options.defaultValue), so you can write your own isDefaultValue() helper if you need to emulate RHF's behavior.`

But there's no explanation of the reason for this choice. But it was apparently unusual enough to warrant its own note in the docs.

If `isDirty` can't, or won't be changed, perhaps there's a way to expose a value indicating that the form is no longer in its default state. This is very useful for disabling submit buttons for example.

For most if not all form libraries I've used, the expected behavior is the form is not dirty when the default values is equal. See formik example here:
https://stackblitz.com/edit/react-hooks-form-validation-example-formik-hh9rdieb

Perhaps it would be appropriate to 
Option 1: Return `isDirty` to it's well known behavior of returning to `false` when the form is at its default values.
Option 2: As option 1, but `isPristine` is no longer a mirror of `isDirty`, but rather `isPristine` will be false when *any* change to the form has been made.
Option 3: Expose a new property, like `isModified` or `isChanged` to indicate that the values differ from the default/initial values.

**Comment by nimo23** [comment]
I really think the current behavior of `isDirty` is a bug - the explanation in the docs doesn't make it any less of a bug. 

I mean, if `isDirty == true`, even if the field values ​​in the end are actually not changed from their original state, then `TanStack form` is giving us wrong information because `isDirty = true` is just wrong. 

Therefore, the current behavior of `isDirty` feels like a bug..

**Comment by ioson** [comment]
i just run into this and think there should be a way to check for a changed form state.
The current implementation is confusing.

Either change isDirty or add isModified or isChanged as alexqhj suggested.

**Reply by przemoch-dev** [reply]
This is great idea. Let's keep track of both the existing isDirty attribute and the new isModified attribute.

**Comment by przemoch-dev** [comment]
Bump!

**Comment by madsh93** [comment]
Yes, the current state does not make sense. isDirty should **never** be true when currentValues === defaultValues. Then it's simply not dirty. It is clean as a whistle.

**Comment by harry-whorlow** [comment]
@alexqhj I'm working on this now, hopefully will have a PR ready by the end of the day 🤟 

How dose this sound `with meta.isDefaultValue`?

**Reply by CanRau** [reply]
Personally I think `meta.isDefaultValue` would be a nice addition, tho `isDirty` should behave as discussed here before, returning to false if `value === defaultValue`.
But maybe I misinterpreted your comment?

**Reply by MartinCura** [reply]
changing the meaning of `isDirty` would be a breaking change, so i'm guessing that's a reason to simply add a new property; `isDefaultValue` sounds fine to me btw

**Reply by CanRau** [reply]
To me it's like a "bug"-fix that possibly warrants a major version bump 🤓 
But this is just me, [this](https://github.com/TanStack/form/discussions/758#discussioncomment-12951805) hook is a good enough workaround currently

**Reply by LeCarbonator** [reply]
> To me it's like a "bug"-fix that possibly warrants a major version bump 🤓 But this is just me, [this](https://github.com/TanStack/form/discussions/758#discussioncomment-12951805) hook is a good enough workaround currently

There's nothing to fix. The field is no longer pristine, even if the value happens to align with the default. The only way to return to the defaultValue is to touch a pristine field and dirtying it, then editing it again so it returns. That's not a pristine field. A separate property is the way to go in my opinion.

[The docs specifically mentioning this behaviour and describing isDirty](<https://tanstack.com/form/latest/docs/framework/react/guides/basic-concepts#field-state>)

[The PR for isDefaultValue](<https://github.com/TanStack/form/pull/1456>)

**Reply by CanRau** [reply]
Hmm yeeaa ok aligning it with pristine I get it now 😅

Would still argue `isPristine` as you are saying `isDirty` like op but I see how that would be confusing 😅

Wanted to suggest `isChanged` that could behave like op wants `isDirty` to behave but it's still ambiguous I guess and `isDefaultValue` is the clearest in the end 👌 

Thanks for making me understand 💛

**Reply by harry-whorlow** [reply]
Good to see the discussion happening 😄

So, @LeCarbonator has made a great point 
> The only way to return to the defaultValue is to touch a pristine field and dirtying it, then editing it again so it returns. That's not a pristine field
---
But, to build on it....

`isDefaultValue` is pretty much the inverse of RHF implementation of isDirty:

| field is dirty  | field value is default |  | RHF isDirty | isDefault |
| --- | ---| --- | --- | --- |
| true  | true  | | false | true |
| false  | true  | | false | true |
| true  | false  | | true | false |

So doing something like `!isDefaultValue` is equivalent to what I'm calling `flippable isDirty™️`, or how most of the react ecosystem handles isDirty.

However when you consider the other frameworks, there's quite a split of implementation’s:

Flippable isDirty (when `value === default` && `dirty === true`, isDirty = false ):
- RHF
- Formic
- Final form

Persistent isDirty (when `value !== default` && `dirty === true`, isDirty =true ):
- Angular form
- Vue FormKit

(_non exhaustive list, please correct me if I'm wrong_ 😊)

We're kind of in a situation where either way we roll this someone won't be happy. But, having isDefaultValue allows us to get the same functionality of RHF (albeit inverse). Whilst also allowing, you the user, to track if the field has been dirtied at any point in its life, whereas in `flippable isDirty™️` this information is lost.

I'm going to rewrite the docs in my PR #1456 so hopefully it's clearer, in the meantime if you have any comments drop them here... I'll additionally ping this issue, once I'm done, so you can add you input to the PR 😄.

**Reply by CanRau** [reply]
A flippable `isDirty` should be accompanied by something like `isChanged` that's always `true` once a field has been changed regardless of it's been reversed to default.

My own arguments where because I was used to the RHF behavior, but I'm fine with both versions as long as you can get all the information

**Comment by harry-whorlow** [comment]
@alexqhj how dose this do? feel free to leave a review over at #1456

![docs](https://github.com/user-attachments/assets/33964201-86bd-439d-93f1-aae011aa5363)

**Comment by harry-whorlow** [comment, marked as answer]
#1456 was just merged giving you access to `meta.isDefaultValue` which if you do something like `!meta.isDefaultValue` will give you the reacts ecosystems isDirty. 

The docs are [here](https://tanstack.com/form/latest/docs/framework/react/guides/basic-concepts#understanding-isdirty-in-different-libraries:~:text=Understanding%20%27isDirty%27%20in%20Different%20Libraries). We'd love to hear some feedback😄

**Reply by jcramb** [reply]
Is it possible to use this in a similar way to isDirty for the entire form?

e.g. in my solid app currently I have `const isDirty = form.useStore(state => state.isDirty);` will I be able to write this with the same approach as `isDefaultValue`? 

I'm really confused on why "persistent dirty" would be the selected approach - what's the recommended approach for disabling a submit button to prevent unneccessary submissions of the same data?

**Reply by harry-whorlow** [reply]
Yeah, you would use `const isDefaultValue = useStore(state => state.isDefaultValue)`

Then if you don't want to be able to submit the same data, your button would look like 
```tsx
<button type="submit" disabled={!canSubmit || isDefaultValue}>
  {isSubmitting ? '...' : 'Submit'}
</button>
```



---

With regards to this:
> I'm really confused on why "persistent dirty" would be the selected approach

- [LeCarbonator](https://github.com/TanStack/form/discussions/1081?notification_referrer_id=NT_kwDOBLmxEbQxMzkwMTk4NDc1ODo3OTI3ODM1Mw#discussioncomment-12978883) comment
- [my](https://github.com/TanStack/form/discussions/1081?notification_referrer_id=NT_kwDOBLmxEbQxMzkwMTk4NDc1ODo3OTI3ODM1Mw#discussioncomment-12984047) comment 
- changing it would be a breaking change... and then the Angular and Vue ecosystem would be upset

Hope this helps/clarifies 😄

**Reply by danvoyce** [reply]
I've achieved this like so

```tsx
<form.Subscribe selector={(state) => state.isDefaultValue}>
  {(isDefaultValue) => (
    <Button type="submit" disabled={isDefaultValue}>
      Save changes
    </Button>
  )}
</form.Subscribe>
```

**Reply by Escapado** [reply]
I just noticed that the `isDefaultValue` does not reset, if we call `form.reset` with new form values.

```ts
form.reset() // isDefaultValue becomes true, isDirty becomes false
form.reset({...newValues}) // same as above, the argument seems to be ignored
form.reset({...newValues}, {keepDefaultValues: true) // Form resets to new values, isDirty becomes false but isDefaultValue stays false
```
Ideally I would like to either see `isDefaultValue` be set to `true` in the third case or have some other API to set it to be `true`. For long lived forms that auto-submit in the background every 10 or so seconds and where you only want to submit if there are truly changes the current behavior is a problem because after the first submit `isDefaultValue` can not be used to check for "real" changes and `isDirty` as we know is not sufficient. So then you have to track the difference between your form state and the last submitted state manually, i.e. you have to compute equality on every change (even though Tanstack does this internally already).
